lib/faraday/request/retry.rb

```diff
@@ -106,26 +112,32 @@ def initialize(app, options = nil)
       @errmatch = build_exception_matcher(@options.exceptions)
     end
 
-    def sleep_amount(retries)
+    def sleep_amount(retries, env)
       retry_index = @options.max - retries
       current_interval = @options.interval * (@options.backoff_factor ** retry_index)
       current_interval = [current_interval, @options.max_interval].min
       random_interval  = rand * @options.interval_randomness.to_f * @options.interval
-      current_interval + random_interval
+
+      retry_after = calculate_retry_after(env)
+      interval    = current_interval + random_interval
+
+      retry_after && retry_after >= interval ? retry_after : interval
     end
 
     def call(env)
       retries = @options.max
       request_body = env[:body]
       begin
         env[:body] = request_body # after failure env[:body] is set to the response body
-        @app.call(env)
+        @app.call(env).tap do |resp|
+          raise Faraday::Error::RetriableResponse.new(nil, resp) if @options.retry_statuses.include?(resp.status)
```

Comment:
I can see what you're trying to do with `Faraday::Error::RetriableResponse`, however I don't think it's a good idea to introduce this exception and use it this way.
The reason is that, after we reach the maximum number of retries, the new exception will be propagated outside and this might be an unexpected behaviour compared to what happens now.

I like the idea of allowing to define a list of statuses that you want to retry, but this should be tested internally without raising any exception.
My suggestion is to add this to `retry_request?`.

Reply:
I've tried to do this without raising an exception but this will lead to implementing two places with retry logic which seems a bit overhead. It could be done using `do ... while` loop but it looks hacky for me.

And I'll move statuses testing to the better place.

Reply:
I've tried to move this line around but it seems that this is the best option.